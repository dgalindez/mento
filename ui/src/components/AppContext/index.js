import { createContext, useEffect } from 'react';

import useGeolocation from '../../hooks/useGeolocation';
import useLocale from '../../hooks/useLocale';
import useWeatherApi from '../../hooks/useWeatherApi';

const AppContext = createContext({
  geolocation: {},
  locale: {},
  weather: {},
});

export const ContextComponent = ({ children }) => {
  const geolocation = useGeolocation();
  const locale = useLocale();
  const weather = useWeatherApi();

  useEffect(() => {
    if (!geolocation?.failed) {
      weather?.getLocationData(geolocation.lat, geolocation.lon);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [geolocation.failed, geolocation.lat, geolocation.lon]);

  return (
    <AppContext.Provider value={{ geolocation, locale, weather }}>
      {children}
    </AppContext.Provider>
  );
};

export default AppContext;
