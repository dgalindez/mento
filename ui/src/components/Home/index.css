.hidden {
  border: none;
}

.home {
  height: 472px;
}
