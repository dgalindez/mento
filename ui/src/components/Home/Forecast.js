import { useContext } from 'react';
import { useTranslation } from 'react-i18next';

import { UNIT_SYMBOLS } from '../../utils/constants';
import AppContext from '../AppContext';

const Forecast = () => {
  const { t } = useTranslation();

  const {
    locale: {
      units,
    },
    weather: {
      daily,
    },
  } = useContext(AppContext);

  return (
    <div className="column">
      <table className="table">
        <thead>
          <tr>
            <th>{t('forecast.date')}</th>
            <th>{t('forecast.max')} / {t('forecast.min')}</th>
          </tr>
        </thead>
        <tbody>
          {daily.map(({ date, max, min }, index) => (
          <tr key={`detail-day-${index}`}>
            <td>{new Date(date).toLocaleDateString()}</td>
            <td>{max} {UNIT_SYMBOLS[units]} / {min} {UNIT_SYMBOLS[units]}</td>
          </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default Forecast;
