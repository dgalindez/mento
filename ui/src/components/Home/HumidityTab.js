import { useContext } from 'react';
import { useTranslation } from 'react-i18next';

import AppContext from '../AppContext';

const TemperatureTab = () => {
  const { t } = useTranslation();
  const {
    weather: {
      current,
    },
  } = useContext(AppContext);

  return (
    <div className="m-4 mx-6">
      {t('home.tab.humidity')} {current.humidity}
    </div>
  );
};

export default TemperatureTab;
