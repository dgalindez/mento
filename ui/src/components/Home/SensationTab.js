import { useContext } from 'react';
import { useTranslation } from 'react-i18next';

import { UNIT_SYMBOLS } from '../../utils/constants';
import AppContext from '../AppContext';

const TemperatureTab = () => {
  const { t } = useTranslation();
  const {
    locale: {
      units,
    },
    weather: {
      current,
    },
  } = useContext(AppContext);

  return (
    <div className="m-4 mx-6">
      {t('home.tab.sensation')} {current.feelsLike} {UNIT_SYMBOLS[units]}
    </div>
  );
};

export default TemperatureTab;
