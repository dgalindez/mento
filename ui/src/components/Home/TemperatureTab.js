import { useContext } from 'react';
import { useTranslation } from 'react-i18next';

import { UNIT_SYMBOLS } from '../../utils/constants';
import AppContext from '../AppContext';

const TemperatureTab = () => {
  const { t } = useTranslation();
  const {
    locale: {
      units,
    },
    weather: {
      current,
    },
  } = useContext(AppContext);

  return (
    <div className="m-4 mx-6">
      {t('home.tab.temperature')} {current.temperature} {UNIT_SYMBOLS[units]}
      <br />
      {t('home.tab.max')} {current.max} {UNIT_SYMBOLS[units]}
      <br />
      {t('home.tab.min')} {current.min} {UNIT_SYMBOLS[units]}
    </div>
  );
};

export default TemperatureTab;
