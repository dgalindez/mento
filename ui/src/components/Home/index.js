import { useContext } from 'react';
import { Link, useLocation } from 'react-router-dom';
import { useTranslation } from 'react-i18next';

import AppContext from '../AppContext';

import Forecast from './Forecast';

import './index.css';

const Home = () => {
  const { t } = useTranslation();
  const {
    weather: {
      city,
      country,
      error,
      loading,
    },
  } = useContext(AppContext);

  const tabs = ['/temperature', '/sensation', '/humidity'];
  const { pathname = '' } = useLocation();

  if (error || loading) {
    return (
      <div className="columns" role="heading" aria-level="1">
        <div className="control is-large is-loading">
          <input className="hidden" />
        </div>
      </div>
    );
  }

  return (
    <>
      <div className="tabs is-boxed is-large mt-4 mx-6" role="heading" aria-level="1">
        <ul>
          {tabs.map((tab, index) => (
            <li className={pathname.includes(tab) ? 'is-active' : ''} key={`tab-${index}`}>
              {/* eslint-disable-next-line jsx-a11y/anchor-is-valid */}
              <Link to={tab}>
                {t(`home.tab.${tab.substring(1)}`)}
              </Link>
            </li>
          ))}
        </ul>
      </div>
      <div className="columns mx-6">
        <div className="column">
          <span className="title">{city}, </span>
          <span className="subtitle">{country}</span>
        </div>
        <Forecast />
      </div>
    </>
  );
};

export default Home;
