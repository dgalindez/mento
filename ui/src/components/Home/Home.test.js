import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { I18nextProvider } from 'react-i18next';

import i18n from '../../utils/i18nTestHelper';
import AppContext from '../AppContext';

import Home from '.';

let mockPathname = '';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({ pathname: mockPathname })
 }));

describe('Home', () => {
  let contextData = {};

  beforeEach(() => {
    contextData = {
      locale: {
        units: 'metric',
      },
      weather: {
        city: 'Seattle',
        country: 'US',
        daily: [],
        error: '',
        loading: false,
      },
    }
  });

  it('renders correctly', () => {
    render(
      <BrowserRouter>
        <I18nextProvider i18n={i18n}>
          <AppContext.Provider value={contextData}>
            <Home />
          </AppContext.Provider>
        </I18nextProvider>
      </BrowserRouter>
    );

    expect(screen.getByRole('heading')).toMatchSnapshot();
  })

  it('renders temperature tab as active', () => {
    mockPathname = '/temperature';
    render(
      <BrowserRouter>
        <I18nextProvider i18n={i18n}>
          <AppContext.Provider value={contextData}>
            <Home />
          </AppContext.Provider>
        </I18nextProvider>
      </BrowserRouter>
    );

    expect(screen.getByRole('heading')).toMatchSnapshot();
  })

  it('renders humidity tab as active', () => {
    mockPathname = '/humidity';
    render(
      <BrowserRouter>
        <I18nextProvider i18n={i18n}>
          <AppContext.Provider value={contextData}>
            <Home />
          </AppContext.Provider>
        </I18nextProvider>
      </BrowserRouter>
    );

    expect(screen.getByRole('heading')).toMatchSnapshot();
  })

  it('renders sensation tab as active', () => {
    mockPathname = '/sensation';
    render(
      <BrowserRouter>
        <I18nextProvider i18n={i18n}>
          <AppContext.Provider value={contextData}>
            <Home />
          </AppContext.Provider>
        </I18nextProvider>
      </BrowserRouter>
    );

    expect(screen.getByRole('heading')).toMatchSnapshot();
  })

  it('renders loading if data is being fetched', () => {
    contextData.weather.loading = true;
    render(
      <I18nextProvider i18n={i18n}>
        <AppContext.Provider value={contextData}>
          <Home />
        </AppContext.Provider>
      </I18nextProvider>
    );

    expect(screen.getByRole('heading')).toMatchSnapshot();
  })

  it('renders loading if fetching data failed', () => {
    contextData.weather.error = 'some error';
    render(
      <I18nextProvider i18n={i18n}>
        <AppContext.Provider value={contextData}>
          <Home />
        </AppContext.Provider>
      </I18nextProvider>
    );

    expect(screen.getByRole('heading')).toMatchSnapshot();
  })
});
