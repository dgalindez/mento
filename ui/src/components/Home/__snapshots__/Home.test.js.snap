// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders correctly 1`] = `
<div
  aria-level="1"
  class="tabs is-boxed is-large mt-4 mx-6"
  role="heading"
>
  <ul>
    <li
      class=""
    >
      <a
        href="/temperature"
      >
        home.tab.temperature
      </a>
    </li>
    <li
      class=""
    >
      <a
        href="/sensation"
      >
        home.tab.sensation
      </a>
    </li>
    <li
      class=""
    >
      <a
        href="/humidity"
      >
        home.tab.humidity
      </a>
    </li>
  </ul>
</div>
`;

exports[`Home renders humidity tab as active 1`] = `
<div
  aria-level="1"
  class="tabs is-boxed is-large mt-4 mx-6"
  role="heading"
>
  <ul>
    <li
      class=""
    >
      <a
        href="/temperature"
      >
        home.tab.temperature
      </a>
    </li>
    <li
      class=""
    >
      <a
        href="/sensation"
      >
        home.tab.sensation
      </a>
    </li>
    <li
      class="is-active"
    >
      <a
        href="/humidity"
      >
        home.tab.humidity
      </a>
    </li>
  </ul>
</div>
`;

exports[`Home renders loading if data is being fetched 1`] = `
<div
  aria-level="1"
  class="columns"
  role="heading"
>
  <div
    class="control is-large is-loading"
  >
    <input
      class="hidden"
    />
  </div>
</div>
`;

exports[`Home renders loading if fetching data failed 1`] = `
<div
  aria-level="1"
  class="columns"
  role="heading"
>
  <div
    class="control is-large is-loading"
  >
    <input
      class="hidden"
    />
  </div>
</div>
`;

exports[`Home renders sensation tab as active 1`] = `
<div
  aria-level="1"
  class="tabs is-boxed is-large mt-4 mx-6"
  role="heading"
>
  <ul>
    <li
      class=""
    >
      <a
        href="/temperature"
      >
        home.tab.temperature
      </a>
    </li>
    <li
      class="is-active"
    >
      <a
        href="/sensation"
      >
        home.tab.sensation
      </a>
    </li>
    <li
      class=""
    >
      <a
        href="/humidity"
      >
        home.tab.humidity
      </a>
    </li>
  </ul>
</div>
`;

exports[`Home renders temperature tab as active 1`] = `
<div
  aria-level="1"
  class="tabs is-boxed is-large mt-4 mx-6"
  role="heading"
>
  <ul>
    <li
      class="is-active"
    >
      <a
        href="/temperature"
      >
        home.tab.temperature
      </a>
    </li>
    <li
      class=""
    >
      <a
        href="/sensation"
      >
        home.tab.sensation
      </a>
    </li>
    <li
      class=""
    >
      <a
        href="/humidity"
      >
        home.tab.humidity
      </a>
    </li>
  </ul>
</div>
`;
